### DISCUSSION TITLE
Possible Issue With Optimization Loop Tutorial/ Developer Documentation

### DISCUSSION THREAD

**Discussion by trq5014** [MPoL-dev/MPoL#66] [Q&A]
When trying to compile and run the opimization.py file, I ran into an issue where I could not convert the gridder object into a pytorch dataset (gridder.to_pytorch_dataset()). I got the following error: 
![image](https://user-images.githubusercontent.com/84112808/120421510-6e3ae000-c334-11eb-8d09-f5445902ac8e.png)

I also get this on jupyter: "To export with uncertainties, first grid visibilities with weighting='uniform', no tapering function, and robust=None. Otherwise, data weights are not defined."

I am using the pip install mpol version instead of the developer version for this, and when I try to graph the visibilities (gridder.get_dirty_image(weighting='uniform')) I get: "get_dirty_image() got an unexpected keyword argument 'weighting'", and therefore when I try to run get_dirty_image() with no arguments I get: "'Gridder' object has no attribute 'C'". However, with the developer version I do not have these issues with get_dirty_image(). 

Furthermore, under the Developer Documentation, I ran into issue while attempting to install the '.[dev]' version of MPoL on windows meriting the following error:
![image](https://user-images.githubusercontent.com/84112808/120422410-2f0d8e80-c336-11eb-8902-1284f6abdd32.png)
If I am doing something wrong, let me know. I am more familiar with working in MacOS's terminal than I am in windows cmd prompt so I may be a little out of my element here. 

I also had an issue with creating the virtual environment as described here. As windows installs the 'activate' file under 'venv\Scripts\activate' and 'source' is not recognized as an internal or external command...

**Comment by hgrzy** [comment]
As far as the virtual environment goes, this has documentation on what command to run in windows to activate it: https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/

**Reply by hgrzy** [reply]
For the '.[dev]', I think you cannot install that unless you get the version from git hub as opposed to pip. The reason I think that is because the developer documentation  says " These packages are only needed when developing, building, and testing the package. They are not needed to run the package in its normal configuration, hence, they are not installed with the vanilla pip install MPoL." But if I'm interpreting that wrong please let me know! Hope that helps!

**Reply by trq5014** [reply]
I see what youre saying, I wasnt clear enough in my explanation. I git cloned my forked MPoL repository and that is where the '.[dev]' is raising that error

**Comment by RCF42** [comment]
I think that the pip install version only gets the 0.1.0 version which is out of date for the tutorials. The way dirty images and weighting is done has changed and the tutorial has the up to date commands but 0.1.0 doesn't.

**Reply by RCF42** [reply]
So I think you just have to use the git install method to get the current version (.1.1.dev)

**Comment by iancze** [comment]
For all of your development efforts, I highly recommend that you install your fork of the project from source in editable mode. 

For example, Tyler would do

```
git clone https://github.com/trq5014/MPoL.git
```

and then create/switch to whatever branch he's using for development. 

After creating and activating a python virtual environment, he would switch to 

```
cd MPoL
pip install -e '.[dev]'
```

to install the package and its development dependencies. The `-e` flag means that any local changes he makes to the source code will be automatically available when he does say. The `.` means "install the package here in this folder," and the [dev] means additionally install the packages that are part of the 'dev' [extra_requires](https://github.com/MPoL-dev/MPoL/blob/main/setup.py#L52).
```
$ python
$ import mpol
```

The installation from [PyPI](https://pypi.org/project/MPoL/) is best for someone who is a user but not developer of the package. So this is normally what happens when you use stable packages like numpy, scipy, etc... Our goal is to periodically release stable packages (I plan to release v0.1.1 when we complete the following [issues](https://github.com/MPoL-dev/MPoL/milestone/1)), but of course these will not reflect the latest development changes that are made after the release.

You can always see which package is installed in your python (virtual) environment by doing 

``pip list``

If you've done the `-e` install correctly, you'll want to see a line that looks like 

``MPoL                          0.1.1.dev0  /Users/ian/Documents/Research/Disks/RML/MPoL/src``

or wherever you have the folder located locally.

If you see something like 

``MPoL                          0.1.0`` 

then you'll need to uninstall this PyPI version (``pip uninstall MPoL``) and then reinstall following the developer documentation (`pip install -e '.[dev]'`).

**Comment by iancze** [comment]
I'm not sure what's going on with Tyler's question about failing `.[dev]` installs. Does it work if you remove the ticks?

```
pip install -e .[dev]
```
?

I agree with your feeling that this might be an issue with the Windows command prompt. I too am quite unfamiliar.

**Reply by RCF42** [reply]
Yeah I think it is just the ticks, I'm using windows and when I use VS Code I can have the tick marks but just going straight from the command prompt it doesn't recognize them.

**Reply by iancze** [reply]
Thanks, that's helpful. Does it work in the windows shell if you use double ticks, e.g. `".[dev]"` ?

The reason the ticks are there in first place is to avoid the behavior of the zsh shell (which is default on Macs now) with regards to [square brackets](https://kinopyo.com/en/blog/escape-square-bracket-by-default-in-zsh). But it's not great that it creates more problems with windows!

**Reply by RCF42** [reply]
It works just fine with double ticks yeah, so just single ticks seems to be a problem

**Reply by trq5014** [reply]
Yes thank you! I was able to install it without the single tick marks. Thank you guys for the help!

**Reply by iancze** [reply]
Glad you found a solution. I'll update the developer docs to use double ticks. It looks like this is what [exoplanet](https://docs.exoplanet.codes/en/stable/user/dev/) recommends as well.
